pkg/provider/conjur/provider.go

```diff
@@ -0,0 +1,150 @@
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package conjur
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"github.com/cyberark/conjur-api-go/conjurapi"
+	"github.com/cyberark/conjur-api-go/conjurapi/authn"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	esv1beta1 "github.com/external-secrets/external-secrets/apis/externalsecrets/v1beta1"
+)
+
+var (
+	errConjurClient          = "cannot setup new Conjur client: %w"
+	errMissingStore          = fmt.Errorf("missing store provider")
+	errMissingConjurProvider = fmt.Errorf("missing store provider Conjur")
+)
+
+type Provider struct {
+	ConjurClient Client
+}
+
+type Client interface {
+	RetrieveSecret(secret string) (result []byte, err error)
+}
+
+func (p *Provider) NewClient(ctx context.Context, store esv1beta1.GenericStore, kube client.Client, namespace string) (esv1beta1.SecretsClient, error) {
+	cfg, err := getProvider(store)
+	if err != nil {
+		return nil, err
+	}
+	config := conjurapi.Config{
+		Account:      *cfg.ServiceAccount,
+		ApplianceURL: *cfg.ServiceURL,
+	}
+	conjur, err := conjurapi.NewClientFromKey(config,
+		authn.LoginPair{
+			Login:  *cfg.ServiceUser,
+			APIKey: *cfg.ServiceAPIKey,
+		},
+	)
+
+	if err != nil {
+		return nil, fmt.Errorf(errConjurClient, err)
+	}
+	p.ConjurClient = conjur
+	return p, nil
+}
+
+func getProvider(store esv1beta1.GenericStore) (*esv1beta1.ConjurProvider, error) {
+	if store == nil {
+		return nil, errMissingStore
+	}
+	spc := store.GetSpec()
+	if spc == nil || spc.Provider == nil || spc.Provider.Conjur == nil {
+		return nil, errMissingConjurProvider
+	}
+	return spc.Provider.Conjur, nil
+}
+
+// Empty GetAllSecrets.
+func (p *Provider) GetAllSecrets(ctx context.Context, ref esv1beta1.ExternalSecretFind) (map[string][]byte, error) {
+	// TO be implemented
+	return nil, fmt.Errorf("GetAllSecrets not implemented")
+}
+
+// GetSecret returns a single secret from the provider.
+func (p *Provider) GetSecret(ctx context.Context, ref esv1beta1.ExternalSecretDataRemoteRef) ([]byte, error) {
+	secretValue, err := p.ConjurClient.RetrieveSecret(ref.Key)
+	if err != nil {
+		return nil, err
+	}
+
+	return secretValue, nil
+}
+
+// GetSecretMap returns multiple k/v pairs from the provider.
+func (p *Provider) GetSecretMap(ctx context.Context, ref esv1beta1.ExternalSecretDataRemoteRef) (map[string][]byte, error) {
+	// Gets a secret as normal, expecting secret value to be a json object
+	data, err := p.GetSecret(ctx, ref)
+	if err != nil {
+		return nil, fmt.Errorf("error getting secret %s: %w", ref.Key, err)
+	}
+
+	// Maps the json data to a string:string map
+	kv := make(map[string]string)
```

Comment:
this does not allow using nested json structures or non-string values like float, int, array, null etc.

We may want to factor out this part here: https://github.com/external-secrets/external-secrets/blob/main/pkg/provider/vault/vault.go#L697-L719

Reply:
All values stored in Conjur are returned as string values. In my testing, this did not negatively effect the operator's ability to provide the secret variable from Conjur to Kubernetes Secrets.  Was there another reason this was suggested?

Reply:
The values are stored as string, right but when we unmarshal them it can break:

Works: simple string key/value pairs:

```json
{
  "foo": "bar"
}
```

This breaks:

```json
{
  "foo": 1234
}
```

And we can observe the error `unable to unmarshal secret BotApp/secretVar: json: cannot unmarshal number into Go value of type string`.

That raises the question: should the provider only support string/string key/value pairs?

If you want to support more you should take a look at [the vault provider implementation](https://github.com/external-secrets/external-secrets/blob/main/pkg/provider/vault/vault.go#L676-L704) which supports all data types and nested json as well. Though that can be totally added at a later stage.